### PR TITLE
Upgrade ES support to version 6.x

### PR DESCRIPTION
**Pull request by jraddaoui**
- Upgrade `elasticsearch` dependency in dashboard and MCPClient.
- Split multi document type indexes to single document indexes:
  - `aips` and `aipfiles` indexes for AIP documents.
  - `transfers` and `transferfiles` indexes for Transfers documents.
- Assume indexes are created correctly and only check existence.
- Avoid the use of `doc_type` and use a default one in all indexes.
- Pass `job.pyprint` to index AIP and files function and normalize
  its use in the index Transfer and files call.
- Improve dashboard commands to rebuild ES indexes.
- Change string fields declaration, use `text` and `keyword`.
- Add `raw` subfields to sort over text fields.
- Check `SEARCH_ENABLED` setting (enhanced in #1172) to create indexes.
- Move indexes check and creation to setup call.
- Create single call to index AIP and AIP files and to index Transfer
  and Transfer files, instead of making it separately.
- Do not check shards after resource deletion.
- Cosmetic changes and reorganization of `elasticSearchFunctions`.
- Reduce max query size to 10000 (default ES limit).
- Add custom analyzer to filenames and paths to create tokens splitting
  on selected chars.
- Properly raise exception in `_try_to index` after max attempts.
- Use `_source` instead of `fields` to get and access the selected
  fields (it doesn't include single values in lists).
- Stop using the `_all` field and checking the mapping to get sub-fields.
- Move filters inside boolean query.
- Improve boolean query generation.
- Use `sum` aggregation to get total AIPs size.
- Install `unar` in dashboard Docker image to allow AIP extraction
  while running the rebuild AIPs indexes command.

**TODO:**

- [x] Modify tests.
- [x] Add task to migrate from 1.x cluster.

**TODO?:**

1. Improve ES hosts declaration:

To allow setting a list of hosts separated by comma.

2. Modify `elasticSearchFunctions` to a class:

To avoid using global variables and passing the client around. The class could be initiated in `wsgi.py` like it's being setup now.

3. Normalize field names to `snake_case` in all indexes:

Field names are quite a mess. I didn't want to change this as external tools may be using the indexes but, since the indexes have changed anyway to be single document indexes, it may be a good moment to do so.

4. Remove ES dependency in dashboard:

Only used to wrap some ES errors, but I think we could find a way to do it via `archivematicaCommon.elasticSearchFunctions` or fully remove it.

5. Avoid index check and creation on the fly:

We could force to run the rebuild indexes commands on the install/upgrade process to avoid checking and creating the indexes on the fly. The check is a HEAD request, so it's not a huge deal, but I think we should not do it.

6. Remove cluster health and index retry systems:

Trust timeout setting and let if fail.

7. Use fixed ids in documents:

Use the UUID of each resource as the ES document `_id` to avoid making queries to get the document ids for deletes and updates.

8. Improve backlog and archival storage searches:

Multiple requests are made on the views. This will require several changes and may not be doable in the time remaining but we should try to not do an extra count request (required by the pager system) and a query over the files indexes while querying later over the `aips` and `transfers` index to show the packages instead fo the files but make the query over the files.

9. Improve `boolean` query generation:

To properly represent the user input in the search form. Explained in [here](https://github.com/artefactual/archivematica/blob/ce471f687a45a0465b1ebc4497baeb32ef7378bc/src/dashboard/src/components/advanced_search.py#L134-L142).

Connects to #1171.
Connects to #1173.
Connects to https://github.com/archivematica/Issues/issues/144.

### CODE REVIEW

**Comment by jraddaoui** [comment]
Hi @sevein (@ross-spencer and @jhsimpson your thoughts are welcome too),

This is ready for a first review. I tried to not change a lot the current implementation in this first round and just focus on upgrading the version support, but I've added some TODOs? in the description for things that could be improved further, and I guess you'll have a few more. We may not have time to do them all, so I'd like to hear you thoughts about them and the changes to this point before moving forward.

Thanks!

File: src/dashboard.Dockerfile [hunk 1]
```diff
@@ -18,6 +18,7 @@ RUN set -ex \
 		nodejs \
 		locales \
 		locales-all \
+		unar \
```
Comment (sevein) [inline src/dashboard.Dockerfile:21]: The same dependency needs to be added to Dashboard's osdeps files.

File: src/archivematicaCommon/lib/elasticSearchFunctions.py [hunk 2]
```diff
@@ -1027,16 +1025,9 @@ def delete_matching_documents(client, index, doc_type, field, value):
...
-    results = client.delete_by_query(
-        index=index,
-        doc_type=doc_type,
-        body=query
-    )
-
+    results = client.delete_by_query(index=index, body=query)
     logger.info('Deleted by query %s', results)
 
-    return results['_indices'][index]['_shards']['successful'] == results['_indices'][index]['_shards']['total']
```
Comment (sevein) [inline src/archivematicaCommon/lib/elasticSearchFunctions.py:1090]: What are the implications of not returning? e.g. `delete_aip_files` is still returning the returned value.

File: src/archivematicaCommon/lib/elasticSearchFunctions.py [hunk 3]
```diff
@@ -112,6 +112,10 @@ class TooManyResultsError(ElasticsearchError):
 _es_hosts = None
 _es_client = None
 DEFAULT_TIMEOUT = 10
+INDICES = ['aips', 'aipfiles', 'transfers', 'transferfiles']
```
Comment (sevein) [inline src/archivematicaCommon/lib/elasticSearchFunctions.py]: Nitpick: indices or indexes? Both are valid I think but we're using both, it'd be nice if we just use one of the forms across the board for consistency.

File: src/archivematicaCommon/lib/elasticSearchFunctions.py [hunk 4]
```diff
@@ -201,21 +205,140 @@ def wait_for_cluster_yellow_status(client, wait_between_tries=10, max_tries=10):
             time.sleep(wait_between_tries)
 
 
-def create_indexes_if_needed(client):
-    create_aips_index(client)
-    create_transfers_index(client)
+def _create_indexes_if_needed(client):
+    """
+    Checks if all indeces exist in the client. Otherwise, creates the missing
```
Comment (sevein) [inline src/archivematicaCommon/lib/elasticSearchFunctions.py]: Typo: `s/indeces/indices`.

File: src/archivematicaCommon/lib/elasticSearchFunctions.py (hunk 3 above)
Comment (sevein) [reply src/archivematicaCommon/lib/elasticSearchFunctions.py]: You could add a comment here describing how the code expects funcs like `_get_aips_index_body` to be defined. It'd be nice for the developer to know what it takes to add a new index.

File: src/archivematicaCommon/lib/elasticSearchFunctions.py [hunk 5]
```diff
@@ -201,21 +205,140 @@ def wait_for_cluster_yellow_status(client, wait_between_tries=10, max_tries=10):
...
+    ones with their settings and mappings.
+    """
```
Comment (sevein) [inline src/archivematicaCommon/lib/elasticSearchFunctions.py:219]: Note that in AM docstrings are written following a different format. See https://www.python.org/dev/peps/pep-0257/#id17:

> Multi-line docstrings consist of a summary line just like a one-line docstring, followed by a blank line, followed by a more elaborate description. The summary line may be used by automatic indexing tools; it is important that it fits on one line and is separated from the rest of the docstring by a blank line. The summary line may be on the same line as the opening quotes or on the next line. The entire docstring is indented the same as the quotes at its first line (see example below).

File: src/archivematicaCommon/lib/elasticSearchFunctions.py [hunk 6]
```diff
@@ -201,21 +205,140 @@ def wait_for_cluster_yellow_status(client, wait_between_tries=10, max_tries=10):
...
+    if client.indices.exists(index=','.join(INDICES)):
+        logger.info('All indices already created.')
+        return
+    for index in INDICES:
+        # Call get index body functions bellow for each index
```
Comment (sevein) [inline src/archivematicaCommon/lib/elasticSearchFunctions.py]: Typo: below.

File: src/dashboard/src/main/management/commands/rebuild_transfer_backlog.py [hunk 7]
```diff
@@ -92,19 +103,22 @@ def prepdir(self, path):
         if tail != suffix:
             return os.path.join(path, suffix)
 
-    def delete_index(self, es_client):
-        """Delete search index."""
-        self.stdout.write('Deleting all transfers in the "transfers" index...')
+    def delete_indexes(self, es_client, indexes):
+        """Delete search indeesx."""
```
Comment (sevein) [inline src/dashboard/src/main/management/commands/rebuild_transfer_backlog.py]: Typo.

File: src/archivematicaCommon/lib/elasticSearchFunctions.py [hunk 8]
```diff
@@ -112,6 +112,10 @@ class TooManyResultsError(ElasticsearchError):
 _es_hosts = None
 _es_client = None
 DEFAULT_TIMEOUT = 10
+INDICES = ['aips', 'aipfiles', 'transfers', 'transferfiles']
+# A doc type is still required in ES 6.x but it's limited to one per index.
+# It will be removed in ES 7.x, so we'll use the same for all indexes.
+DOC_TYPE = '_doc'
```
Comment (sevein) [inline src/archivematicaCommon/lib/elasticSearchFunctions.py:113]: Makes sense!

File: src/archivematicaCommon/lib/elasticSearchFunctions.py [hunk 9]
```diff
@@ -201,21 +205,140 @@ def wait_for_cluster_yellow_status(client, wait_between_tries=10, max_tries=10):
...
+    Checks if all indeces exist in the client. Otherwise, creates the missing
+    ones with their settings and mappings.
+    """
+    if client.indices.exists(index=','.join(INDICES)):
+        logger.info('All indices already created.')
+        return
+    for index in INDICES:
+        # Call get index body functions bellow for each index
+        body = getattr(sys.modules[__name__], '_get_%s_index_body' % index)()
+        logger.info('Creating "%s" index ...' % index)
```
Comment (sevein) [inline src/archivematicaCommon/lib/elasticSearchFunctions.py]: String replacement in `logging` is usually done as follows:

    logger.info('Creating "%s" index ...', index)

File: src/archivematicaCommon/lib/elasticSearchFunctions.py [hunk 10]
```diff
@@ -381,16 +409,16 @@ def index_aip(client, uuid, name, filePath, pathToMETS, size=None, aips_in_aic=N
     dublincore = root.find('mets:dmdSec/mets:mdWrap/mets:xmlData/dcterms:dublincore', namespaces=ns.NSMAP)
     if dublincore is not None:
         aip_type = dublincore.findtext('dc:type', namespaces=ns.NSMAP) or dublincore.findtext('dcterms:type', namespaces=ns.NSMAP)
-        if aip_type == "Archival Information Collection":
+        if aip_type == 'Archival Information Collection':
```
Comment (sevein) [inline src/archivematicaCommon/lib/elasticSearchFunctions.py:457]: JD used to prefer double quoting. I think we need to write down our style preference project-wise so we don't keep making the same kind of change.

File: src/dashboard/src/components/advanced_search.py [hunk 11]
```diff
@@ -199,8 +197,9 @@ def filter_search_fields(es_client, search_fields, index=None, doc_type=None):
             continue
         try:
             field_name = field.rsplit('.', 1)[0]
-            mapping = elasticSearchFunctions.get_type_mapping(es_client, index, doc_type)
-            subfields = mapping[doc_type]['properties'][field_name]['properties']
+            # FIXME
```
Comment (sevein) [inline src/dashboard/src/components/advanced_search.py]: ?

File: src/dashboard/src/components/archival_storage/views.py [hunk 12]
```diff
@@ -412,12 +410,13 @@ def elasticsearch_query_excluding_aips_pending_deletion(uuid_field_name):
...
 def aip_file_count(es_client):
     query = elasticsearch_query_excluding_aips_pending_deletion('AIPUUID')
-    return advanced_search.indexed_count(es_client, 'aips', ['aipfile'], query)
+    return advanced_search.indexed_count(es_client, 'aipfiles', query)
 
 
 def total_size_of_aips(es_client):
     query = elasticsearch_query_excluding_aips_pending_deletion('uuid')
     query['fields'] = 'size'
+    # FIXME
```
Comment (sevein) [inline src/dashboard/src/components/archival_storage/views.py]: No facets, is that the problem?

File: src/archivematicaCommon/lib/elasticSearchFunctions.py [hunk 13]
```diff
@@ -851,7 +851,7 @@ def search_all_results(client, body, index='', **query_params):
         body=body,
         index=index,
         size=MAX_QUERY_SIZE,
-        **query_params)
+    )
```
Comment (sevein) [inline src/archivematicaCommon/lib/elasticSearchFunctions.py:952]: Nice!

File: src/dashboard/src/components/backlog/views.py [hunk 14]
```diff
@@ -174,6 +173,7 @@ def search(request):
             sort=order_by + ':' + sort_direction if order_by else '',
             _source=source,
         )
+        hit_count = hits['hits']['total']
```
Comment (sevein) [inline src/dashboard/src/components/backlog/views.py:181]: Nice!

**Review by sevein — COMMENTED** [review]
Looking good, @jraddaoui. I hope the tests can be updated easily!

Your TODO lists some good ideas. Hopefully we can address those separately, we can probably find good ways to group them where it makes sense.

**Comment by sevein** [comment]
I've rebased!

File: src/dashboard/src/components/advanced_search.py (hunk 11 above)
Comment (jraddaoui) [reply src/dashboard/src/components/advanced_search.py]: Mapping changes for this version required some fixes in this code. In the end with ES 6.x there was no need to get the sub-fields, check https://github.com/artefactual/archivematica/pull/1285/commits/7018be000089c95911b1888bdf02abc8424f0cbd.

File: src/dashboard/src/components/archival_storage/views.py (hunk 12 above)
Comment (jraddaoui) [reply src/dashboard/src/components/archival_storage/views.py]: Yes, it was fixed in https://github.com/artefactual/archivematica/pull/1285/commits/f7ba54379a2132d3063304b58400554446c846b5.

File: src/archivematicaCommon/lib/elasticSearchFunctions.py (hunk 10 above)
Comment (jraddaoui) [reply src/archivematicaCommon/lib/elasticSearchFunctions.py:457]: OK, sorry. I'll try to change quotes as less as possible until we have the style defined.

File: src/archivematicaCommon/lib/elasticSearchFunctions.py (hunk 2 above)
Comment (jraddaoui) [reply src/archivematicaCommon/lib/elasticSearchFunctions.py:1090]: The returned value is not checked in any of the delete calls. I have removed some of those not needed returns in https://github.com/artefactual/archivematica/pull/1285/commits/1129d8f068dfa355527a2f827eeba80212af0c25.

File: src/dashboard.Dockerfile (hunk 1 above)
Comment (jraddaoui) [reply src/dashboard.Dockerfile:21]: Done.

File: src/archivematicaCommon/lib/elasticSearchFunctions.py (hunk 3 above)
Comment (jraddaoui) [reply src/archivematicaCommon/lib/elasticSearchFunctions.py]: Done.

File: src/archivematicaCommon/lib/elasticSearchFunctions.py (hunk 5 above)
Comment (jraddaoui) [reply src/archivematicaCommon/lib/elasticSearchFunctions.py:219]: Thanks! I have fixed a few more in https://github.com/artefactual/archivematica/pull/1285/commits/28a56030181f6edd3646de0c6433b76ce81dd018.

**Comment by jraddaoui** [comment]
Hi @sevein, this is ready for another round! Sorry for the delay.

**Comment by jraddaoui** [comment]
I have created https://github.com/archivematica/Issues/issues/404 for the other improvements from the description.

File: src/archivematicaCommon/tests/test_elasticsearch_functions.py [hunk 15]
```diff
@@ -6,6 +6,11 @@
 
 import elasticSearchFunctions
 
+try:
+    from unittest.mock import ANY, patch
+except ImportError:
+    from mock import ANY, patch
```
Comment (sevein) [inline src/archivematicaCommon/tests/test_elasticsearch_functions.py:13]: I've been just using `from mock import ANY, patch` assuming that it becomes no-op in Python3 but I guess you way is more explicit.

**Review by sevein — APPROVED** [review]
Thanks, @jraddaoui! Once this is merged, have you thought about next action items? I can imagine this is going to require some work in other repos, e.g.:

- https://github.com/artefactual-labs/ansible-archivematica-src (are we going to need a new branch?)
- https://github.com/artefactual/archivematica-docs (branch 1.9?)
- https://github.com/artefactual/deploy-pub (playbooks)

**Comment by jraddaoui** [comment]
Good point @sevein, I've sent an email to the team to check a few things before merging ;)

File: src/dashboard/src/main/management/commands/reindex_from_remote_cluster.py [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+        body = {
+            "source": {
+                "remote": {
+                    "host": options["host"],
+                    # "socket_timeout": options["socket_timeout"],
+                    # "connect_timeout": options["connect_timeout"],
+                },
+                "index": "",
+                "type": "",
+                # "size": options["size"],
```
Comment (mamedin) [inline src/dashboard/src/main/management/commands/reindex_from_remote_cluster.py]: Hi Radda,

I tested a migration from ES version 1.7.6 to version 6.5.4. The aips index has 24805 documents, and the data directory is 177M. I found this error when running the new `reindex_from_remote_cluster` command:

```
archivematica@mamedin-elasticcsearch-test:/usr/share/archivematica/dashboard$ /usr/share/archivematica/virtualenvs/archivematica-dashboard/bin/python manage.py reindex_from_remote_cluster http://127.0.0.1:9600
Reindexing aips:
Traceback (most recent call last):
  File "manage.py", line 10, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/share/archivematica/virtualenvs/archivematica-dashboard/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 354, in execute_from_command_line
    utility.execute()
  File "/usr/share/archivematica/virtualenvs/archivematica-dashboard/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 346, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/share/archivematica/virtualenvs/archivematica-dashboard/local/lib/python2.7/site-packages/django/core/management/base.py", line 394, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/share/archivematica/virtualenvs/archivematica-dashboard/local/lib/python2.7/site-packages/django/core/management/base.py", line 445, in execute
    output = self.handle(*args, **options)
  File "/opt/archivematica/archivematica/src/dashboard/src/main/management/commands/reindex_from_remote_cluster.py", line 147, in handle
    body=body,
  File "/usr/share/archivematica/virtualenvs/archivematica-dashboard/local/lib/python2.7/site-packages/elasticsearch/client/utils.py", line 76, in _wrapped
    return func(*args, params=params, **kwargs)
  File "/usr/share/archivematica/virtualenvs/archivematica-dashboard/local/lib/python2.7/site-packages/elasticsearch/client/__init__.py", line 783, in reindex
    params=params, body=body)
  File "/usr/share/archivematica/virtualenvs/archivematica-dashboard/local/lib/python2.7/site-packages/elasticsearch/transport.py", line 318, in perform_request
    status, headers_response, data = connection.perform_request(method, url, params, body, headers=headers, ignore=ignore, timeout=timeout)
  File "/usr/share/archivematica/virtualenvs/archivematica-dashboard/local/lib/python2.7/site-packages/elasticsearch/connection/http_urllib3.py", line 186, in perform_request
    self._raise_error(response.status, raw_data)
  File "/usr/share/archivematica/virtualenvs/archivematica-dashboard/local/lib/python2.7/site-packages/elasticsearch/connection/base.py", line 125, in _raise_error
    raise HTTP_EXCEPTIONS.get(status_code, TransportError)(status_code, error_message, additional_info)
elasticsearch.exceptions.RequestError: RequestError(400, u'illegal_argument_exception', u'Remote responded with a chunk that was too large. Use a smaller batch size.')
```
I tested with several batch sizes, for example 50 and 10. It works for size=10:

```
root@mamedin-elasticcsearch-test:/opt/archivematica/archivematica/src/dashboard# git diff src/main/management/commands/reindex_from_remote_cluster.py
diff --git a/src/dashboard/src/main/management/commands/reindex_from_remote_cluster.py b/src/dashboard/src/main/management/commands/reindex_from_remote_cluster.py
index 2184cd97..12a1b024 100644
--- a/src/dashboard/src/main/management/commands/reindex_from_remote_cluster.py
+++ b/src/dashboard/src/main/management/commands/reindex_from_remote_cluster.py
@@ -99,6 +99,7 @@ class Command(DashboardCommand):
                 },
                 "index": "",
                 "type": "",
+                "size": "10"
                 # "size": options["size"],
             },
             "dest": {
```

It took 47s to reindex, this is the output including the `time`command:

```artefactual@mamedin-elasticcsearch-test:~/elasticsearch-1.7.6/elasticsearch-1.7.6$ sudo su - archivematica -s /bin/bash
archivematica@mamedin-elasticcsearch-test:~$ export $(cat /etc/default/archivematica-dashboard)
-su: export: `#': not a valid identifier
archivematica@mamedin-elasticcsearch-test:~$ cd /usr/share/archivematica/dashboard
archivematica@mamedin-elasticcsearch-test:/usr/share/archivematica/dashboard$ ARCHIVEMATICA_DASHBOARD_DASHBOARD_ELASTICSEARCH_TIMEOUT=300 time /usr/share/archivematica/virtualenvs/archivematica-dashboard/bin/python manage.py reindex_from_remote_cluster http://127.0.0.1:9600
Reindexing aips:
{
    "retries": {
        "bulk": 0, 
        "search": 0
    }, 
    "updated": 0, 
    "batches": 3, 
    "throttled_until_millis": 0, 
    "throttled_millis": 0, 
    "noops": 0, 
    "created": 115, 
    "deleted": 0, 
    "took": 15592, 
    "requests_per_second": -1.0, 
    "timed_out": false, 
    "failures": [], 
    "version_conflicts": 0, 
    "total": 115
}
Reindexing aipfiles:
{
    "retries": {
        "bulk": 0, 
        "search": 0
    }, 
    "updated": 0, 
    "batches": 506, 
    "throttled_until_millis": 0, 
    "throttled_millis": 0, 
    "noops": 0, 
    "created": 24690, 
    "deleted": 0, 
    "took": 30185, 
    "requests_per_second": -1.0, 
    "timed_out": false, 
    "failures": [], 
    "version_conflicts": 0, 
    "total": 24690
}
Reindexing transfers:
{
    "retries": {
        "bulk": 0, 
        "search": 0
    }, 
    "updated": 0, 
    "batches": 0, 
    "throttled_until_millis": 0, 
    "throttled_millis": 0, 
    "noops": 0, 
    "created": 0, 
    "deleted": 0, 
    "took": 11, 
    "requests_per_second": -1.0, 
    "timed_out": false, 
    "failures": [], 
    "version_conflicts": 0, 
    "total": 0
}
Reindexing transferfiles:
{
    "retries": {
        "bulk": 0, 
        "search": 0
    }, 
    "updated": 0, 
    "batches": 0, 
    "throttled_until_millis": 0, 
    "throttled_millis": 0, 
    "noops": 0, 
    "created": 0, 
    "deleted": 0, 
    "took": 10, 
    "requests_per_second": -1.0, 
    "timed_out": false, 
    "failures": [], 
    "version_conflicts": 0, 
    "total": 0
}
0.60user 0.10system 0:47.25elapsed 1%CPU (0avgtext+0avgdata 57308maxresident)k
4440inputs+48outputs (0major+24062minor)pagefaults 0swaps
```

**Review by mamedin — CHANGES_REQUESTED** [review]

**Comment by jraddaoui** [comment]
Hi @sevein,

I have squashed all the work already reviewed and added a new commit which includes a command to re-index the documents from a remote cluster. I'd appreciate if you could review that commit ;)

@mamedin, I have added a couple of optional arguments to the command to allow setting the chunk size and the request timeout. I'll follow-up about that in the Redmine ticket and related email thread.

**Review by mamedin — APPROVED** [review]
I tested again with the new options for batch size and timeout and it works fine.

**Comment by jraddaoui** [comment]
Hi @sevein,

I have simplified the command a bit as I was adding options to keep and merge the indexes that in the end I think we should remove, to avoid confusions and ease the documentation. Also, the `--timeout` (`-t`) param will now be used for both connections to avoid changing the config/env only for the execution of this command.

**Review by sevein — APPROVED** [review]
@jraddaoui the command looks fine, thanks! I haven't tested but no red flags. One thought is that you subclass `DashboardCommand` but you're not really taking advantage of its methods. You certainly could. In particular I wonder if we should control exceptions raised by `reindex` so we can print friendlier messages.

:+1:

**Comment by jraddaoui** [comment]
> @jraddaoui the command looks fine, thanks! I haven't tested but no red flags. One thought is that you subclass `DashboardCommand` but you're not really taking advantage of its methods. You certainly could. In particular I wonder if we should control exceptions raised by `reindex` so we can print friendlier messages.

Thanks @sevein! I've improved a little the output and error handling of the command. Please, let me know if that's okay before I merge.

**Review by sevein — APPROVED** [review]
Cool, thanks! :+1:

**Pull request merged**